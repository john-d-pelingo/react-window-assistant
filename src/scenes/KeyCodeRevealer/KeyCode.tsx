import React from 'react'
import styled from 'styled-components'

const KeyCodeContainer = styled.div`
  display: table-cell;
  vertical-align: middle;
`

const KeyCodeNumber = styled.span`
  display: block;
  font-size: 300px;
  line-height: 320px;
  padding-bottom: 40px;
`

const KeyCodeButton = styled.button`
  background: #eff0f2;
  border-radius: 4px;
  border-top: 1px solid rgba(255, 255, 255, 0.8);
  box-shadow: inset 0 0 25px #e8e8e8, 0 1px 0 #c3c3c3, 0 2px 0 #c9c9c9,
    0 2px 3px #333;
  color: #aaa;
  cursor: pointer;
  display: inline-block;
  font: 38px/40px 'Helvetica Neue';
  font-weight: bold;
  margin: 0;
  padding: 15px 30px;
  position: relative;
  text-align: center;
  text-decoration: none;
  text-shadow: 0 1px 0 rgba(255, 255, 255, 0.8);
`

interface IKeyCodeProps {
  handleClick?: (event: React.MouseEvent<HTMLSpanElement>) => void
  keyCode: number
  keyText?: string
}

export const KeyCode: React.FC<IKeyCodeProps> = ({
  keyCode,
  keyText,
  handleClick = () => undefined,
}) => (
  <KeyCodeContainer>
    <KeyCodeNumber aria-label="key-code-number">
      {isNaN(keyCode) ? '!=Num' : keyCode}
    </KeyCodeNumber>
    <KeyCodeButton aria-label="key-code-button" onClick={handleClick}>
      {keyText || 'What key code is that?'}
    </KeyCodeButton>
  </KeyCodeContainer>
)
