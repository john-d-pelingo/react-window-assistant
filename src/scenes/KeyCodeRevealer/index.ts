export { KeyCodeRevealer as default } from './KeyCodeRevealer'
