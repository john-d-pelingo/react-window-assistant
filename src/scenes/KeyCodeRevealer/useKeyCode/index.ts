import { produce } from 'immer'
import { useEffect, useReducer, useRef } from 'react'

import { keyCodes } from '../keyCodes'
import { blur, focus, resetKeyCode, setNewKeyCode } from './actions'
import {
  BLUR,
  FOCUS,
  IKeyCodeState,
  IPotentialKeyCodes,
  KeyCodeActionTypes,
  RESET_KEY_CODE,
  SET_NEW_KEY_CODE,
} from './types'

function reducer(
  state: IKeyCodeState,
  action: KeyCodeActionTypes,
): IKeyCodeState {
  return produce(state, draft => {
    switch (action.type) {
      case BLUR:
        draft.isBlurred = true
        break

      case RESET_KEY_CODE:
        draft.newKeyCode = null
        break

      case SET_NEW_KEY_CODE:
        draft.newKey = action.payload.key
        draft.newKeyCode = action.payload.keyCode
        break

      case FOCUS:
        draft.isBlurred = false
        break
    }
  })
}

export function useKeyCode<T extends HTMLElement>(initialState: IKeyCodeState) {
  const appElement = useRef<T>(null)

  const [{ isBlurred, newKey, newKeyCode }, dispatch] = useReducer(
    reducer,
    initialState,
  )

  const handleBlur = () => {
    if (!isBlurred) {
      dispatch(blur())
    }
  }

  const handleFocus = () => {
    if (isBlurred) {
      dispatch(focus())
    }
  }

  const handleClick = () => {
    dispatch(resetKeyCode())
  }

  const handleKeyDown = (potentialKeyCode: IPotentialKeyCodes) => {
    const keyCode = potentialKeyCode.which
    const key = keyCodes[keyCode]
      ? keyCodes[keyCode]
      : potentialKeyCode.key.toLowerCase()

    dispatch(
      setNewKeyCode({
        key,
        keyCode: String(keyCode),
      }),
    )
  }

  const focusAppElement = () => {
    if (appElement.current) {
      appElement.current.focus()
    }
  }

  useEffect(() => {
    if (isBlurred) {
      focusAppElement()
    }
  }, [isBlurred])

  return {
    appElement,
    handleBlur,
    handleClick,
    handleFocus,
    handleKeyDown,
    newKey,
    newKeyCode,
  }
}
