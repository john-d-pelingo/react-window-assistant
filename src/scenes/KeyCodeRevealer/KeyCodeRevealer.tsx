import React, { FC, useEffect } from 'react'
import { Helmet } from 'react-helmet'
import styled from 'styled-components'

import { KeyCode } from './KeyCode'
import { KeyCodeState, useKeyCode } from './useKeyCode'

const KeyCodeRevealerContainer = styled.div`
  display: table;
  height: 60vh;
  margin: 0 auto;
  text-align: center;
  width: 100%;
  outline: none;
`

const DefaultText = styled.span`
  display: table-cell;
  font-size: 50px;
  line-height: 55px;
  vertical-align: middle;
`

const documentTitle = 'Key Code Revealer'

const initialState: KeyCodeState = {
  isBlurred: true,
  newKey: '',
  newKeyCode: null,
}

// TODO: connect with query parameter to make sharable
export const KeyCodeRevealer: FC = () => {
  const {
    appElement,
    handleBlur,
    handleClick: handleKeyCodeClick,
    handleFocus,
    handleKeyDown: handleKeyCodeKeyDown,
    newKey,
    newKeyCode,
  } = useKeyCode<HTMLDivElement>(initialState)

  useEffect(() => {
    if (newKey && newKeyCode) {
      document.title = `${newKeyCode} : ${newKey}`
    }
  }, [newKey, newKeyCode])

  const handleClick = (event: React.MouseEvent<HTMLSpanElement>) => {
    event.preventDefault()
    handleKeyCodeClick()
  }

  const handleKeyDown = (event: React.KeyboardEvent<HTMLDivElement>) => {
    // Keep the original synthetic event around.
    // See https://fb.me/react-event-pooling for more information.
    event.persist()

    if (!event.metaKey) {
      event.preventDefault()
    }

    handleKeyCodeKeyDown({
      key: event.key,
      which: event.which,
    })
  }

  return (
    <>
      <Helmet>
        <title>{documentTitle}</title>
      </Helmet>
      <KeyCodeRevealerContainer
        aria-label="key-code-app"
        tabIndex={0}
        onBlur={handleBlur}
        onFocus={handleFocus}
        // TODO: maybe attach to window
        onKeyDown={handleKeyDown}
        ref={appElement}
      >
        {newKeyCode ? (
          <KeyCode
            keyCode={Number(newKeyCode)}
            keyText={newKey}
            handleClick={event => {
              handleClick(event)
              document.title = documentTitle
            }}
          />
        ) : (
          <DefaultText>Press a key in your keyboard</DefaultText>
        )}
      </KeyCodeRevealerContainer>
    </>
  )
}
