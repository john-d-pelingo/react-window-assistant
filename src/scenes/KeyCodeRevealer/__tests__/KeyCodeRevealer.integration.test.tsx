import '@testing-library/jest-dom/extend-expect'
import 'jest-styled-components'

import {
  fireEvent,
  getByLabelText as rootGetByLabelText,
  render,
} from '@testing-library/react'
import { createMemoryHistory } from 'history'
import React from 'react'
import { Helmet } from 'react-helmet'
import { Router } from 'react-router-dom'

import { KeyCodeRevealer } from '../KeyCodeRevealer'

let memoryHistory = createMemoryHistory()

describe('scenes - KeyCodeRevealer', () => {
  beforeEach(() => {
    memoryHistory = createMemoryHistory()
  })

  it('mounts', () => {
    expect(memoryHistory.location.pathname).toBe('/')

    const { container } = render(
      <Router history={memoryHistory}>
        <KeyCodeRevealer />
      </Router>,
    )

    expect(Helmet.peek().title).toBe('Key Code Revealer')
    expect(container.firstChild).toMatchSnapshot()
  })

  it('shows the key pressed and changes document title to the corresponding key on key press', () => {
    expect(memoryHistory.location.pathname).toBe('/')

    const { getByLabelText } = render(
      <Router history={memoryHistory}>
        <KeyCodeRevealer />
      </Router>,
    )

    const appElement = getByLabelText('key-code-app') as HTMLDivElement

    fireEvent.keyDown(appElement, {
      keyCode: 13,
    })

    const keyCodeNumberElement = getByLabelText(
      'key-code-number',
    ) as HTMLSpanElement
    const keyCodeButtonElement = getByLabelText(
      'key-code-button',
    ) as HTMLSpanElement

    expect(keyCodeNumberElement.textContent).toBe('13')
    expect(keyCodeButtonElement.textContent).toBe('enter')
    expect(Helmet.peek().title).toBe('13 : enter')
    expect(memoryHistory.location.search).toBe('?key=13')

    fireEvent.keyDown(appElement, {
      keyCode: 187,
    })

    expect(keyCodeNumberElement.textContent).toBe('187')
    expect(keyCodeButtonElement.textContent).toBe('equal sign')
    expect(Helmet.peek().title).toBe('187 : equal sign')
    expect(memoryHistory.location.search).toBe('?key=187')
  })

  it('shows the key pressed and changes the document title to an unknown key on an unknown key press', () => {
    expect(memoryHistory.location.pathname).toBe('/')

    const { getByLabelText } = render(
      <Router history={memoryHistory}>
        <KeyCodeRevealer />
      </Router>,
    )

    const appElement = getByLabelText('key-code-app')

    fireEvent.keyDown(appElement, {
      keyCode: 999,
    })

    const keyCodeNumberElement = getByLabelText(
      'key-code-number',
    ) as HTMLSpanElement
    const keyCodeButtonElement = getByLabelText(
      'key-code-button',
    ) as HTMLSpanElement

    expect(keyCodeNumberElement.textContent).toBe('999')
    expect(keyCodeButtonElement.textContent).toBe('unidentified')
    expect(Helmet.peek().title).toBe('999 : unidentified')
    expect(memoryHistory.location.search).toBe('?key=999')
  })

  it('resets view on key press and click of the button', () => {
    expect(memoryHistory.location.pathname).toBe('/')

    const { getByLabelText } = render(
      <Router history={memoryHistory}>
        <KeyCodeRevealer />
      </Router>,
    )

    const appElement = getByLabelText('key-code-app')

    expect(appElement.textContent).toMatch(/press a key/i)

    fireEvent.keyDown(appElement, {
      keyCode: 18,
    })

    expect(appElement.textContent).toMatch(/^\d/)
    expect(appElement.textContent).toMatch(/alt/i)
    expect(appElement.textContent).not.toMatch(/press a key/i)

    const keyCodeButtonElement = rootGetByLabelText(
      appElement,
      'key-code-button',
    )

    fireEvent.click(keyCodeButtonElement)

    expect(appElement.textContent).toMatch(/press a key/i)
    expect(Helmet.peek().title).toBe('Key Code Revealer')
    expect(memoryHistory.location.search).toBe('?')
  })

  it('loads a correct key from the URL', () => {
    expect(memoryHistory.location.pathname).toBe('/')

    memoryHistory.push('?key=82')

    const { getByLabelText } = render(
      <Router history={memoryHistory}>
        <KeyCodeRevealer />
      </Router>,
    )

    expect(getByLabelText(/key-code-button/i)).toHaveTextContent('r')
  })

  it('loads a wrong key from the URL', () => {
    expect(memoryHistory.location.pathname).toBe('/')

    memoryHistory.push('?key=fake-key-code')

    const { getByLabelText } = render(
      <Router history={memoryHistory}>
        <KeyCodeRevealer />
      </Router>,
    )

    expect(getByLabelText(/key-code-number/i)).toHaveTextContent('!=Num')
    expect(getByLabelText(/key-code-button/i)).toHaveTextContent(
      'What key code is that?',
    )
    expect(memoryHistory.location.search).toBe('?key=NaN')
  })
})
