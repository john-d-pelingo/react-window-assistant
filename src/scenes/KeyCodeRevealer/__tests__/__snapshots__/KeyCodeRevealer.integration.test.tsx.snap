// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`scenes - KeyCodeRevealer mounts 1`] = `
.c0 {
  display: table;
  height: 60vh;
  margin: 0 auto;
  text-align: center;
  width: 100%;
  outline: none;
}

.c1 {
  display: table-cell;
  font-size: 50px;
  line-height: 55px;
  vertical-align: middle;
}

<div
  aria-label="key-code-app"
  class="c0"
  tabindex="0"
>
  <span
    class="c1"
  >
    Press a key in your keyboard
  </span>
</div>
`;
